### PR TITLE
remove `express-github-webhook`

### LINKED ISSUE #48: Remove `express-github-webhook` dependency
[`express-github-webhook`](https://github.com/Gisonrg/express-github-webhook) has been a great starting-point for this project. However, its simplicity has begun adding complexity as this project grows. And, as it continues to expand, the problem will get worse. Since `express-github-webhook` is simply a middleware wrapper to the GitHub events API which we could implement ourselves, we should do so.

### PR DESCRIPTION
**Pull request by omh1280**
Fixes #48 

This is a very similar copy of [`express-github-webhook`](https://github.com/Gisonrg/express-github-webhook). So much so that I want to include something saying it is based off of it. Where is the best place to put this disclaimer?

I've ported all the tests to Jasmine, and made sure that they all pass. This is nearly completed, I am going to do a once-over soon and bump when it is ready for a review!

### CODE REVIEW

**Comment by cesium-concierge** [comment]
@omh1280 thanks for the pull request!

I noticed that (`CHANGES.md`)[https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CHANGES.md] has not been updated. If this change updates the Cesium API in any way, fixes a bug, or makes any non-trivial update, please add a bullet point to `CHANGES.md` and bump this pull request so we know it was updated. For more info, see the [Pull Request Guidelines]( https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines).

__I am a bot who helps you make Cesium awesome!__ Thanks again.

File: lib/crontab [hunk 1]
```diff
@@ -0,0 +1 @@
+* * * * * root echo "Hello world" >> /var/log/cron.log 2>&1
```
Comment (mramato) [inline lib/crontab]: I assume submitting this was an accident?

File: package.json [hunk 2]
```diff
@@ -46,6 +46,7 @@
     "jasmine-spec-reporter": "^4.1.1",
     "nyc": "^11.0.3",
     "open": "^0.0.5",
+    "supertest": "^3.0.0",
```
Comment (mramato) [inline package.json]: Why was this added?

**Comment by mramato** [comment]
Can you please describe what is different here from just using `express-github-webhook`?  The goal of not depending on a third party library was so that we can have a streamlined implementation that does exactly what we need.  This looks almost identical to what the library was already doing without any architectural improvements.

**Comment by omh1280** [comment]
Not much--It's essentially the same except it's emitting the repository's `full_name` instead of `name`. This way we can generate API URL's easily from the `full_name`.

Maybe we should meet about this because I'm not sure how this could be improved architecturally.

**Comment by cesium-concierge** [comment]
Thank you for the pull request!

Looks like this pull request hasn't been updated in 0 days since I last commented.
To keep things tidy should this be closed? Perhaps keep the branch and submit an issue?

__I am a bot who helps facilitate your development!__ Have a nice day.

**Comment by omh1280** [comment]
@mramato can you let me know how `lib/checkWebHook` looks? The tests are failing right now, but I'll fix them after you quickly review it.

If you want me to use the function as "middle-ware", it's going to require wrapping the handle with a function since it needs to know `secret` within the middleware.

**Comment by mramato** [comment]
> since it needs to know secret within the middleware.

Can't it just use `Settings.secret`?

**Comment by omh1280** [comment]
Ready for a look!

**Comment by cesium-concierge** [comment]
Thank you for the pull request!

Looks like this pull request hasn't been updated in 0 days since I last commented.
To keep things tidy should this be closed? Perhaps keep the branch and submit an issue?

__I am a bot who helps facilitate your development!__ Have a nice day.

File: lib/checkWebHook.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    Check.typeOf.object('req', req);
+
+    if (req.method !== 'POST') {
+        throw new Error('Request was not POST');
+    }
+
+    // check header fields
+    var id = req.headers['x-github-delivery'];
+    if (!defined(id)) {
+        throw new Error('No id found in the request');
```
Comment (mramato) [inline lib/checkWebHook.js]: All of these errors should be in the form of 

```js
next(new Error('No id found in the request'));
return;
```

In general, don't throw from route handlers.

File: lib/checkWebHook.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+var Check = Cesium.Check;
+var defined = Cesium.defined;
+
+var Settings = require('./Settings');
+module.exports = checkWebHook;
+
+function checkWebHook(req, res, next) { // eslint-disable-line no-unused-vars
+    Check.typeOf.object('req', req);
+
+    if (req.method !== 'POST') {
```
Comment (mramato) [inline lib/checkWebHook.js]: I'm not sure why this check is needed, if the method isn't a post, none of our handlers should be called (assuming the handler is properly set up to only listen to POST).

File: index.js [hunk 5]
```diff
@@ -16,60 +16,67 @@ var Settings = require('./lib/Settings');
...
-    app.use(webHookHandler);
+    app.use(checkWebHook);
 
-    repositoryNames.forEach(function (repositoryName) {
-        dateLog('Listening to ' + repositoryName);
-        webHookHandler.on(repositoryName, function (event, jsonResponse) {
-            dateLog('Received event to repository: ' + repositoryName);
-            dateLog('event: ' + event);
-            dateLog('jsonResponse: ' + jsonResponse);
+    app.use(Settings.listenPath, function (req, res) {
```
Comment (mramato) [inline index.js]: Shouldn't this be `app.post`?  Isn't POST all we care about?

File: index.js [hunk 6]
```diff
@@ -16,60 +16,67 @@ var Settings = require('./lib/Settings');
...
+        var repositoryName = req.body.repository.full_name;
+        var event = req.headers['x-github-event'];
+        if (!(repositoryName in repositoryNames)) {
+            dateLog('Could not find ' + repositoryName + ' in ' + repositoryNames);
+            return;
```
Comment (mramato) [inline index.js]: Shouldn't this use `next(error)` to make sure we return an error?

**Comment by mramato** [comment]
Please run the auto-formatter on all of the new code.

**Comment by omh1280** [comment]
Webstorm's auto-formatter?

**Comment by mramato** [comment]
> Webstorm's auto-formatter?

Yes, CTRL-ALT-L if you're on Windows with the default keymappings, or you can select the files and then do `Code -> Reformat code` from the menu.

**Comment by omh1280** [comment]
Done!

File: lib/checkWebHook.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+var defined = Cesium.defined;
+
+var Settings = require('./Settings');
+module.exports = checkWebHook;
+
+function checkWebHook(req, res, next) { // eslint-disable-line no-unused-vars
+    // check header fields
+    var id = req.headers['x-github-delivery'];
+    if (!defined(id)) {
+        next(new Error('No id found in the request'));
```
Comment (mramato) [inline lib/checkWebHook.js]: You need to `return` after calling `next` in each of these cases

**Comment by omh1280** [comment]
Sorry about that. Fixed

File: index.js [hunk 8]
```diff
@@ -2,82 +2,91 @@
...
+                next(new Error('Got an error: ' + e));
             });
         });
-    });
 
-    webHookHandler.on('error', function (err, req, res) { // eslint-disable-line no-unused-vars
-        dateLog('WebHookHandler got error: ' + err);
-    });
+        // Handle errors
+        app.use(function (err, req, res, next) { // eslint-disable-line no-unused-vars
```
Comment (mramato) [inline index.js]: Are we sure we need this custom error handler?

Comment (omh1280) [inline index.js]: If I remove it, Express automatically returns an error code?

File: index.js [hunk 9]
```diff
@@ -2,82 +2,91 @@
...
+                });
+
+                if (!defined(result)) {
                     dateLog('GitHub request did not match any events the server is listening for');
                     return;
                 }
-                dateLog('GitHub API returned with statusCode: ' + res.statusCode);
-                dateLog('and statusMessage: ' + res.statusMessage);
+                dateLog('GitHub API returned with statusCode: ' + result.statusCode);
+                dateLog('and statusMessage: ' + result.statusMessage);
```
Comment (mramato) [inline index.js]: You should be calling `next` at the end here.

File: index.js [hunk 10]
```diff
@@ -2,82 +2,91 @@
...
+        app.use(bodyParser.json());
+        app.use(checkWebHook);
+
+        app.post(Settings.listenPath, function (req, res, next) {
+            var repositoryName = req.body.repository.full_name;
+            var event = req.headers['x-github-event'];
+            if (!(repositoryName in repositoryNames)) {
+                var message = 'Could not find ' + repositoryName + ' in ' + repositoryNames;
+                dateLog(message);
+                next(new Error(message));
```
Comment (mramato) [inline index.js]: Missing a return after this next.

File: index.js [hunk 11]
```diff
@@ -2,82 +2,91 @@
...
+
+        var app = express();
 
-    var app = express();
-    app.use(bodyParser.json());
-    app.use(webHookHandler);
+        app.use(bodyParser.json());
+        app.use(checkWebHook);
+
+        app.post(Settings.listenPath, function (req, res, next) {
```
Comment (mramato) [inline index.js]: You you please declare this function in a separate file and unit test it.

File: index.js (hunk 8 above)
Comment (mramato) [inline index.js]: As far as I know it does, but don't take my word for it, test it.

File: package.json [hunk 12]
```diff
@@ -26,10 +26,10 @@
   },
   "dependencies": {
     "bluebird": "^3.0.0",
+    "buffer-equal-constant-time": "^1.0.1",
```
Comment (mramato) [inline package.json]: Buffer has a built in `equals` method, no need for this library: https://nodejs.org/api/buffer.html#buffer_buf_equals_otherbuffer

File: lib/checkWebHook.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    next();
+}
+
+checkWebHook.signData = function (secret, data) {
+    return 'sha1=' + crypto.createHmac('sha1', secret).update(data).digest('hex');
+};
+
+checkWebHook.verifySignature = function (secret, data, signature) {
+    return bufferEq(Buffer.from(signature), Buffer.from(checkWebHook.signData(secret, data)));
```
Comment (mramato) [inline lib/checkWebHook.js]: I'm confused, aren't `signature` and the result of `checkWebHook.signData` both strings?  Why are we creating buffers from them simply to compare?

File: lib/checkWebHook.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    var sign = req.headers['x-hub-signature'] || '';
+    if (Settings.secret && !sign) {
+        next(new Error('No signature found in the request'));
+        return;
+    }
+
+    if (!defined(req.body)) {
+        next(new Error('Make sure body-parser is used'));
```
Comment (mramato) [inline lib/checkWebHook.js]: This error message is misleading.  If a post body is not defined, it's not because we're no using body-parser, it's because whomever issued the POST didn't send one.  I would just put something like "Expected req.body to be defined").

File: index.js [hunk 15]
```diff
@@ -2,82 +2,91 @@
...
-    });
+    .then(function (repositoryNames) {
+        dateLog('Loaded settings successfully');
+
+        var app = express();
 
-    var app = express();
-    app.use(bodyParser.json());
-    app.use(webHookHandler);
+        app.use(bodyParser.json());
```
Comment (mramato) [inline index.js]: `app.use` means a function runs on every route, not just routes we care about.  Rather passing `bodyParser.json` and `checkWebHook` to `app.use`, pass them as parameters to the `post` call below, this way they only run on the routes we actually care about.

File: index.js (hunk 9 above)
Comment (omh1280) [inline index.js]: I don't know how Express works very deeply, but does `res.status(200).send()` end the chain of middle-ware?

Comment (mramato) [inline index.js]: In most cases, yes, but there are edge cases where not calling `next` can be an issue, so it's just a good practice to always call next.

**Comment by omh1280** [comment]
Just committed--It's more express-y now.

Full coverage and unit tests pass. Let me know how it looks!

File: index.js [hunk 16]
```diff
@@ -1,83 +1,32 @@
...
-                    dateLog('GitHub request did not match any events the server is listening for');
-                    return;
-                }
-                dateLog('GitHub API returned with statusCode: ' + res.statusCode);
-                dateLog('and statusMessage: ' + res.statusMessage);
-            }).catch(function (e) {
-                dateLog('Got an error: ' + e);
-            });
+        // Handle errors
+        app.use(function (err, req, res, next) { // eslint-disable-line no-unused-vars
```
Comment (mramato) [inline index.js]: So what did you find out about error handling?  Is this needed?

Comment (omh1280) [inline index.js]: seems OK to remove.

When run in production mode, however, the errors won't have a stack trace.

Comment (mramato) [inline index.js]: That's fine.

**Comment by omh1280** [comment]
Seems to work! #50 #51
